internal/worker/apiaddresssetter/worker.go

```diff
@@ -0,0 +1,366 @@
+	}
+
+	// Consume the initial events from the watchers. The watcher will
+	// dispatch an initial event when it is created, so we need to consume
+	// that event before we can start watching.
+	if _, err := eventsource.ConsumeInitialEvent[[]string](ctx, watcher); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return watcher.Changes(), nil
+}
+
+// updateControllerNodes updates the current list of tracked controller nodes,
+// as well as starting and stopping trackers for them as they are added and
+// removed.
+func (w *apiAddressSetterWorker) updateControllerNodes(ctx context.Context) (bool, error) {
+	controllerIDs, err := w.config.ControllerNodeService.GetControllerIDs(ctx)
+	if err != nil {
+		return false, errors.Errorf("cannot get controller IDs: %w", err)
+	}
+
+	w.config.Logger.Debugf(ctx, "controller nodes: %#v", controllerIDs)
+	changed := false
+
+	// Stop controller tracker that no longer correspond to controller nodes.
+	for controllerID := range w.controllerTrackers {
+		if !slices.Contains(controllerIDs, controllerID) {
+			if err:=w.stopAndRemoveTracker(ctx, controllerID):err!=nil{
+				return false, errors.Capture(err)
+			}
+			changed = true
+		}
+	}
+
+	// Start trackers for new nodes.
+	for _, controllerID := range controllerIDs {
+		if _, ok := w.controllerTrackers[controllerID]; ok {
+			continue
+		}
+		w.config.Logger.Debugf(ctx, "found new controller %q", controllerID)
+		tracker, err := newControllerTracker(controllerID, w.config.ApplicationService, w.controllerNodeChanges)
+		if err != nil {
+			return false, errors.Capture(err)
+		}
+		if err := w.runner.StartWorker(controllerID, func() (worker.Worker, error) {
+			return tracker, nil
+		}); err != nil {
+			return false, errors.Errorf("failed to start tracker for controller node %q: %w", controllerID, err)
+		}
+
+		w.controllerTrackers[controllerID] = tracker
+		changed = true
+	}
+
+	return changed, nil
+}
+
+func (w *apiAddressSetterWorker) stopAndRemoveTracker(ctx context.Context, controllerID string) error {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*5)
+	defer cancel()
+
+	// Stop tracker since it's no longer required.
+	w.config.Logger.Debugf(ctx, "stopping tracker for controller node %q", controllerID)
+	if err := w.runner.StopAndRemoveWorker(controllerID, ctx.Done()); errors.Is(err, context.DeadlineExceeded) {
+		return errors.Errorf("failed to stop tracker for controller node %q: timed out", controllerID)
+	} else if err != nil {
+		return errors.Errorf("failed to stop tracker for controller node %q: %w", controllerID, err)
+	}
+	// Remove the tracker from the internal list as well.
+	delete(w.controllerTrackers, controllerID)
+
+	return nil
+}
+
+// updateAPIAddresses updates the API addresses for each tracked controller.
+func (w *apiAddressSetterWorker) updateAPIAddresses(ctx context.Context) error {
+	for controllerID := range w.controllerTrackers {
+		unitName := unit.Name(application.ControllerApplicationName + "/" + controllerID)
```

Comment:
Please use `unit.NewNameFromParts`

Reply:
Or at least `NewName` if you don't want to deal with type conversions, so we can be sure this hand made unit name is valid

Reply:
That's done.